File: core/domain/src/commonMain/kotlin/com/syrmos/core/domain/journey/JourneyDetail.kt
package com.syrmos.core.domain.journey

import com.syrmos.core.model.journey.JourneyOption
import com.syrmos.core.model.journey.Leg
import com.syrmos.core.model.journey.LegKind
import kotlinx.datetime.Instant

/**
 * Selected-journey detail timeline (S05), the Kotlin peer of web
 * `SyrmosJourneyDetail.timeline` and iOS `JourneyDetail.timeline`. Pure transform:
 * a [JourneyOption]'s legs become an ordered list of [TimelineRow]s the S05 view
 * renders. Language-neutral (ids/instants/counts only; the view localizes) and
 * honest about time (a null clock stays null, never a fabricated 0). Matches the
 * shared golden fixture fixtures/journeys/detail.json.
 */
object JourneyDetail {
    /** A node's role on the rail, driving its size/outline in the view. */
    enum class Node { ORIGIN, DESTINATION, INTERCHANGE }

    /** One rendered timeline row. Fields are populated per [kind]. */
    data class TimelineRow(
        val kind: String, // board | stops | alight | transfer | walk
        val legId: String,
        val stationId: String? = null,
        val lineId: String? = null,
        val towardsId: String? = null,
        val clock: Instant? = null,
        val timingKind: String? = null,
        val node: Node? = null,
        val count: Int? = null,
        val fromId: String? = null,
        val toId: String? = null,
        val seconds: Int? = null,
    )

    fun timeline(option: JourneyOption): List<TimelineRow> {
        val legs = option.legs
        val lastIndex = legs.size - 1
        val rows = mutableListOf<TimelineRow>()

        legs.forEachIndexed { i, leg ->
            if (leg.kind == LegKind.RIDE) {
                val stops = leg.orderedStopIds
                val towardsId = stops.lastOrNull() ?: leg.toId
                rows += TimelineRow(
                    kind = "board", legId = leg.id, stationId = leg.fromId, lineId = leg.lineId,
                    towardsId = towardsId, clock = leg.departureInstant,
                    timingKind = leg.timingKind.wire(),
                    node = if (i == 0) Node.ORIGIN else Node.INTERCHANGE,
                )
                val count = maxOf(0, stops.size - 2)
                if (count > 0) rows += TimelineRow(kind = "stops", legId = leg.id, lineId = leg.lineId, count = count)
                rows += TimelineRow(
                    kind = "alight", legId = leg.id, stationId = leg.toId,
                    clock = leg.arrivalInstant, timingKind = leg.timingKind.wire(),
                    node = if (i == lastIndex) Node.DESTINATION else Node.INTERCHANGE,
                )
            } else {
                val prev = legs.getOrNull(i - 1)
                val next = legs.getOrNull(i + 1)
                val seconds = leg.transferMinimumSeconds
                    ?: gapSeconds(prev?.arrivalInstant, next?.departureInstant)
                rows += TimelineRow(
                    kind = if (leg.kind == LegKind.WALK) "walk" else "transfer", legId = leg.id,
                    fromId = leg.fromId, toId = leg.toId, seconds = seconds, node = Node.INTERCHANGE,
                )
            }
        }
        return rows
    }

    private fun gapSeconds(from: Instant?, to: Instant?): Int? {
        if (from == null || to == null) return null
        return (to - from).inWholeSeconds.toInt()
    }

    // TimingKind wire name (lowercase, matches @SerialName and the fixture).
    private fun com.syrmos.core.model.journey.TimingKind.wire(): String = name.lowercase()
}

File: iosApp/iosApp/Core/Journey/JourneyDetail.swift
import Foundation

/// Selected-journey detail timeline (S05), the Swift peer of web
/// `SyrmosJourneyDetail.timeline` and Kotlin `JourneyDetail.timeline`. Pure
/// transform: an ordered list of legs becomes an ordered list of timeline rows the
/// S05 detail view renders. Language-neutral (ids/dates/counts only; the view
/// localizes) and honest about time (a nil clock stays nil, never a fabricated 0).
/// Matches the shared golden fixture fixtures/journeys/detail.json.
enum JourneyDetail {
    /// A node's role on the rail, driving its size/outline in the view.
    enum Node: String, Equatable { case origin, destination, interchange }

    /// One leg fed to the transform (a ride, or a transfer/walk between rides).
    struct DetailLeg: Equatable {
        let id: String
        let kind: String            // "ride" | "transfer" | "walk"
        let lineId: String?
        let fromId: String
        let toId: String
        let orderedStopIds: [String]
        let departure: Date?
        let arrival: Date?
        let transferMinimumSeconds: Int?
        let timingKind: String      // "scheduled" | "estimated" | "live" | "unknown"

        init(id: String, kind: String, lineId: String? = nil, fromId: String, toId: String,
             orderedStopIds: [String] = [], departure: Date? = nil, arrival: Date? = nil,
             transferMinimumSeconds: Int? = nil, timingKind: String = "unknown") {
            self.id = id; self.kind = kind; self.lineId = lineId; self.fromId = fromId; self.toId = toId
            self.orderedStopIds = orderedStopIds; self.departure = departure; self.arrival = arrival
            self.transferMinimumSeconds = transferMinimumSeconds; self.timingKind = timingKind
        }
    }

    /// One rendered timeline row. Fields are populated per `kind`.
    struct TimelineRow: Equatable {
        let kind: String            // board | stops | alight | transfer | walk
        let legId: String
        var stationId: String? = nil
        var lineId: String? = nil
        var towardsId: String? = nil
        var clock: Date? = nil
        var timingKind: String? = nil
        var node: Node? = nil
        var count: Int? = nil
        var fromId: String? = nil
        var toId: String? = nil
        var seconds: Int? = nil
    }

    static func timeline(_ legs: [DetailLeg]) -> [TimelineRow] {
        var rows: [TimelineRow] = []
        let lastIndex = legs.count - 1
        for (i, leg) in legs.enumerated() {
            if leg.kind == "ride" {
                let towardsId = leg.orderedStopIds.last ?? leg.toId
                rows.append(TimelineRow(
                    kind: "board", legId: leg.id, stationId: leg.fromId, lineId: leg.lineId,
                    towardsId: towardsId, clock: leg.departure, timingKind: leg.timingKind,
                    node: i == 0 ? .origin : .interchange))
                let count = max(0, leg.orderedStopIds.count - 2)
                if count > 0 {
                    rows.append(TimelineRow(kind: "stops", legId: leg.id, lineId: leg.lineId, count: count))
                }
                rows.append(TimelineRow(
                    kind: "alight", legId: leg.id, stationId: leg.toId, clock: leg.arrival,
                    timingKind: leg.timingKind, node: i == lastIndex ? .destination : .interchange))
            } else {
                let prev = i > 0 ? legs[i - 1] : nil
                let next = i < legs.count - 1 ? legs[i + 1] : nil
                var seconds = leg.transferMinimumSeconds
                if seconds == nil, let a = prev?.arrival, let b = next?.departure {
                    seconds = Int(b.timeIntervalSince(a).rounded())
                }
                rows.append(TimelineRow(
                    kind: leg.kind == "walk" ? "walk" : "transfer", legId: leg.id,
                    node: .interchange, fromId: leg.fromId, toId: leg.toId, seconds: seconds))
            }
        }
        return rows
    }
}

/// Connection risk (S07 / Phase R), the Swift peer of web `SyrmosConnectionRisk`
/// and Kotlin `ConnectionRisk`. Pure rule for one transfer: given the real gap
/// available, the change time to allow, and any uncertainty, return the honest
/// status. Reuses the shared feasibility margin EXACTLY so the S07 warning cannot
/// drift from the option's feasibility chip (margin = available - recommended -
/// uncertainty; < 0 missed, 0..179 tight, else comfortable; a nil gap -> unknown).
/// Mirrors fixtures/journeys/connection-risk.json.
enum ConnectionRisk {
    static let tightMaxSeconds = 179
    static let defaultRecommendedSeconds = 120

    /// Status string ("comfortable" | "tight" | "missed" | "unknown").
    static func status(availableSeconds: Int?, recommendedSeconds: Int, uncertaintySeconds: Int = 0) -> String {
        guard let available = availableSeconds else { return "unknown" }
        let margin = available - recommendedSeconds - uncertaintySeconds
        if margin < 0 { return "missed" }
        if margin <= tightMaxSeconds { return "tight" }
        return "comfortable"
    }
}
